## The $d$-invariant

### Question 1

Before writing the function to calculate $d(S(p,q),i)$, there are a few things to deal with.
The first is that python will not return an exact fraction for a recurring decimal; intead it will output it as a decimal rounded to 17 d.p. The standard Python library has a function for this:

In [3]:
from fractions import Fraction

def decimal_to_fraction(decimal_number): 
    return Fraction(decimal_number).limit_denominator() #Fraction().limit_denominator finds rational approximation for given floating point number
## returns Fraction(a,b) = a/b

We also need to check if $p,q$ are coprime. This is equivalent to having $gcd(p,q)=1$, so the `gcd` function is useful: 

In [4]:
def gcd(a, b):
    # Inputs: integers a, b"
    # Returns the greatest common divisor of a, b"
    if not(a % 1 == 0  and b % 1 == 0):
        print("Only integer inputs permitted.")
        return None
    if b == 0: 
        return abs(a)
    else:
        return gcd(b, a % b)

Now the $d$-invariant function can be written. It will be recursive, so we will use memoization for it.

In [5]:
mem=dict()
## dictionary for memoization

def d(p, q, i):
    global mem
    if p==1:
        mem[p,q,i] = 0 # a base case
    if gcd(p,q) != 1:
        return None # not a valid input
    else:
        if (p, q, i) not in mem:
        ## main part of the function
            if p>q>0 and 0<=i<(p+q):
                mem[p,q,i] = ((2*i+1 - p - q)**2 - p*q)/(4*p*q) - d(q,p,i) # when conditions are satisfied it runs recursively
            else:
                mem[p,q,i] = d(p,q%p,i%p) # exchanging q and i with q%p and i%p (which is equivalent but now satisfies the above conditions) 
        return decimal_to_fraction(mem[p,q,i]) # returns an exact fraction instead of rounding

To check this runs as intended, we can compare its output for $S(p,q)=S(9,7)$ to the given values.

In [6]:
print(d(9,7,0),d(9,7,1),d(9,7,2),d(9,7,3), d(9,7,4), d(9,7,5),d(9,7,6),d(9,7,7),d(9,7,8))

0 -2/9 4/9 0 4/9 -2/9 0 -8/9 -8/9


Hence the function is working as it should.

In [14]:
d(4,5,3)

Fraction(0, 1)

### Question 2 (core)

Th

In [10]:
def dS(p,q):
    return [d(p,q,i ) for i in range(0,p)]

In [6]:
dS(6,4)

[None, None, None, None, None, None]

### Question 3

We know that $S(p,q)=S(p,q')$ if either $q=q'$ or $qq' \equiv 1 (mod \ p)$.

So we need to compare the $d$-invariants of numbers modulo p and that of their inverse.

In [46]:
## Let's try d(S(9,7)) first
print(dS(9,7))
## 4*7 = 28 = 1 (mod 9)
print(dS(9,4))

[Fraction(0, 1), Fraction(-2, 9), Fraction(4, 9), Fraction(0, 1), Fraction(4, 9), Fraction(-2, 9), Fraction(0, 1), Fraction(-8, 9), Fraction(-8, 9)]
[Fraction(0, 1), Fraction(4, 9), Fraction(4, 9), Fraction(0, 1), Fraction(-8, 9), Fraction(-2, 9), Fraction(0, 1), Fraction(-2, 9), Fraction(-8, 9)]


In [32]:
## Trying with p = 9 again:
print(dS(9,5))
print(dS(9,2))

[Fraction(8, 9), Fraction(2, 9), Fraction(0, 1), Fraction(2, 9), Fraction(8, 9), Fraction(0, 1), Fraction(-4, 9), Fraction(-4, 9), Fraction(0, 1)]
[Fraction(8, 9), Fraction(8, 9), Fraction(0, 1), Fraction(2, 9), Fraction(-4, 9), Fraction(0, 1), Fraction(-4, 9), Fraction(2, 9), Fraction(0, 1)]


In [45]:
## Seeing a pattern with which elements stay the same
## Switching p
print(dS(13,5))
print(dS(13,8))

[Fraction(6, 13), Fraction(8, 13), Fraction(0, 1), Fraction(8, 13), Fraction(6, 13), Fraction(-6, 13), Fraction(-2, 13), Fraction(-8, 13), Fraction(2, 13), Fraction(2, 13), Fraction(-8, 13), Fraction(-2, 13), Fraction(-6, 13)]
[Fraction(6, 13), Fraction(2, 13), Fraction(8, 13), Fraction(-2, 13), Fraction(-2, 13), Fraction(8, 13), Fraction(2, 13), Fraction(6, 13), Fraction(-6, 13), Fraction(-8, 13), Fraction(0, 1), Fraction(-8, 13), Fraction(-6, 13)]


From this it can be seen that the $d$-invariants of $S(p,q)$ and $S(p,q')$ have the same elements, but are in a different order. The exceptions to this are the first and last elements in them. We want to check if this is a pattern true in any case, so we will try to verify for multiple random cases:

In [38]:
from random import randrange # this function selects a random integer in a given range

In [44]:
for i in range(0,10): # 10 cases
    p = randrange(2,50) # select p randomly. Ignores the case p=1 as there is no inverse modulo 1 
    q = randrange(1,p) # select q (mod p)
    if gcd(p,q) == 1: # only runs if there exists an inverse of q modulo p 
        q1 = pow(q,-1,p)
        if dS(p,q)[0] != dS(p,q1)[0] or dS(p,q)[p-1] != dS(p,q1)[p-1]:
            ## compares the value of the first and last elements of each d-invariant
            print("ERROR")

Therefore this suggests that the $d$-invariants of $S(p,q)=S(p,q')$ are lists of the same values - they share the same first and last elements, but those inbetween are shuffled.

### Question 5

Let us call the set of numbers $\frac{m^2}{q}$ that have this property __M__. Firstly we want the list of odd numbers $m<105$ to test for a given $q$.

In [7]:
m = []
for i in range(0,105):
    if i%2 == 1:
        m.append(i)
    else:
        None

The following function outputs the set of fractions $\frac{m^2}{q}$ such that the knot $S(m^2,q)$ has $m$ values of the $d$-invariants that are equal to 0. The value for $q$ is the input - this is done as a way of splitting up the set of numbers that satisfy the given conditions.

In [8]:
def d_zeros(q):
    zeros = [] # begins as empty list
    for i in range(0, len(m)):
        inv = []
        for j in range(0, m[i]**2): # the length of the d-invariant list for S(m^2,q), where m = m[i] each loop
            if d((m[i])**2, q,j) == 0: 
                inv.append(1)
                ## this records each time that one of the d-invariants are zero
        if len(inv) == m[i]:
            zeros.append(decimal_to_fraction(m[i]**2/q))
            ## if the number of zeros == m, then we want to record m^2/q
    return zeros

We should briefly examine if this function is giving the intended output:

In [9]:
d_zeros(4)

[Fraction(1, 4), Fraction(9, 4), Fraction(25, 4)]

In [17]:
print(dS(1,4))
print()
print(dS(9,4))
print()
print(dS(25,4))

[0]

[Fraction(0, 1), Fraction(4, 9), Fraction(4, 9), Fraction(0, 1), Fraction(-8, 9), Fraction(-2, 9), Fraction(0, 1), Fraction(-2, 9), Fraction(-8, 9)]

[Fraction(24, 25), Fraction(36, 25), Fraction(36, 25), Fraction(24, 25), Fraction(0, 1), Fraction(14, 25), Fraction(16, 25), Fraction(6, 25), Fraction(-16, 25), Fraction(0, 1), Fraction(4, 25), Fraction(-4, 25), Fraction(-24, 25), Fraction(-6, 25), Fraction(0, 1), Fraction(-6, 25), Fraction(-24, 25), Fraction(-4, 25), Fraction(4, 25), Fraction(0, 1), Fraction(-16, 25), Fraction(6, 25), Fraction(16, 25), Fraction(14, 25), Fraction(0, 1)]


It can be seen that, for $q=4$, there is 1 $d$-invariant equal to 0 for $m^2=1$, 3 for $m^2=9$ and 5 for $m^2=25$. This indicates that the outputs are valid; but cannot confirm that it every possible fraction is given.

In order to compare our list(s) to it, we should define a function that outputs the set __R__ in [1, Definition 1.1] for a given $q$ and with $p=m^2$ for odd $m<105$.

The set is defined as follows:

"*__R__ contains the set of rational numbers $\frac{p}{q}$ such that $p > q > 0$, $gcd(p, q) = 1$, $p = m^2$ for some $m ∈ N$ and $q$ is of one of the following types:*

*(1) $mk ± 1$ with $m > k > 0$ and $gcd(m, k) = 1$;*

*(2) $d(m ± 1)$, where $d > 1$ divides $2m ∓ 1$;*

*(3) $d(m ± 1)$, where $d > 1$ is odd and divides $m ± 1$.*"

In [22]:
def R_list(q): # again we split up the set based on what q is - this can be used for direct comparison with M
    R = []
    for i in range(0,len(m)): # going through each possible value of p
        if gcd(m[i]**2,q) ==1 and (m[i]**2)>q>0:
        ## initial conditions, then sorted by if q is any of the three types described above
            for k in range(1,m[i]): # (1)
                if k*m[i] +1 ==q or k*m[i] - 1 == q:
                    if gcd(m[i],k) ==1:
                        R.append(decimal_to_fraction(m[i]**2/q))
            
            for j in range(2,208): # (2)
                ## 207 is the greatest value that 2m+1 can be, as max(m)=103
                if j*(m[i]+1)==q and (2*m[i]-1)%j == 0:
                    R.append(decimal_to_fraction(m[i]**2/q))
                if j*(m[i]-1)==q and (2*m[i]+1)%j == 0:
                    R.append(decimal_to_fraction(m[i]**2/q))
            
            for l in range(2,208): # (3)
                if l%2 == 1:
                    if (m[i]+1)%l == 0 and l*(m[i]+1)==q:
                        R.append(decimal_to_fraction(m[i]**2/q))
                    if (m[i]-1)%l == 0 and l*(m[i]-1)==q:
                        R.append(decimal_to_fraction(m[i]**2/q))
    return R

Now we can compare the two sets in a few examples:

In [21]:
print("R contains " +str(R_list(2)))
print("M contains " +str(d_zeros(2)))
print()
print("R contains " +str(R_list(100)))
print("M contains " +str(d_zeros(100)))

R contains [Fraction(9, 2)]
M contains [Fraction(1, 2), Fraction(9, 2)]

R contains [Fraction(121, 100), Fraction(361, 100), Fraction(441, 100), Fraction(9801, 100), Fraction(10201, 100)]
M contains [Fraction(1, 100), Fraction(81, 100), Fraction(121, 100), Fraction(361, 100), Fraction(441, 100), Fraction(9801, 100), Fraction(10201, 100)]


It is clear that as $d(S(1,q),i)=0$ for any $q,i$, the fraction $\frac{1}{q}$ will be part of the set __M__. However, __R__ will not include any fraction with numerator 1; as $p$ must be greater than $q$ for its conditions to hold. Furthermore this means that any other potential fractions that are in __M__ might not be in __R__ due to $p$ being smaller than $q$.

However these are not the only numbers missing from __R__. Consider when $q=37$:

In [13]:
print(R_list(37))
print(d_zeros(37))

[Fraction(81, 37), Fraction(361, 37)]
[Fraction(1, 37), Fraction(81, 37), Fraction(121, 37), Fraction(361, 37)]


$\frac{121}{37}$ is not in __R__ even though it is greater than 1. 

Therefore we can at most say that __R__ is a subset of __M__. To test this we will randomly generate a value of $q$ several times.

In [43]:
for i in range(0,5): # 5 tests
    q = randrange(1,100) # randomly pick q and get the lists
    r=R_list(q)
    z=d_zeros(q)
    for k in range(0,len(r)):
        if r[k] not in z:
            print("ERROR")
            ## all elements in R should be in M - no ERROR messages indicates this

Thus we conclude that all elements in __R__ satisfy the same properties as those in __M__, hence __R__$\subseteq$__M__.

## References:

[1] : P. Lisca, Lens spaces, rational balls and the ribbon conjecture, Geom. Topol. 11 (2007) 429–472. (Also at https://arxiv.org/abs/math/0701610.)